app.py

```diff
@@ -707,6 +759,163 @@ def edit():
+	assigned to the terminal nodes.
+	"""
+	# guardrails against producing illict tree structures
+	tree_copy = tree.copy(deep=True)
+	# if initial parse labels non-gaps as GAP, change to N-Head by default
+	# if initial parse labels punctuation as something other than punctuation, change to proper punctuation category
+	# if initial parse labels non-punctuation as punctuation, change to N-Head
+	for subt in tree_copy.subtrees(lambda t: t.height() == 2):
+		i = subt[0]
+		if subt.label.startswith('GAP') and senttok[i] != '_.':
+			subt.label = 'N-Head'
+		# ensure that label is a punctuation label for punctuation nodes. exception: symbols that might not be punctuation.
+		if is_possible_punct_token(senttok[i]) and senttok[i] not in AMBIG_SYM:
+			for e in PUNCT_ESCAPING:
+				if senttok[i] == e['ptree_token']:
+					subt.label = e['ptree_label']
+					break
+			if senttok[i] not in [e['ptree_token'] for e in PUNCT_ESCAPING]:
+				subt.label = senttok[i]+"-p"
+		if (not is_possible_punct_token(senttok[i])) and (is_punct_label(subt.label) or is_punct_label(subt.label + "-p")):
+			subt.label = 'N-Head'
+
+	# create three lists of equal lengths: one list non-punctuation token strings, one list of lists prepending punctuation, and one list of lists for appending punctuation
+	non_punct_tokens = []
+	prepunct_tokens = [[] for subt in tree_copy.subtrees(lambda t: t.height() == 2) if (not is_punct_label(subt.label))]
+	postpunct_tokens = copy.deepcopy(prepunct_tokens)
+
+	token_counter = 0
+
+	# iterate through the tree to update the three lists simultaneously
+	for subt in tree_copy.subtrees(lambda t: t.height() == 2):
+		i = subt[0]
+		if subt.label in INITIAL_PUNCT_LABEL or (is_punct_label(subt.label) and token_counter == 0):
+			prepunct_tokens[token_counter].append(senttok[i])
+		elif not is_punct_label(subt.label):
+			non_punct_tokens.append(senttok[i])
+			token_counter += 1
+		elif is_punct_label(subt.label) or token_counter == len(senttok) - 1:
+			postpunct_tokens[token_counter - 1].append(senttok[i])
+
+	tree_to_cgel = remove_punctuation_nodes(tree_copy) 
+	
+	try:
+		block = writetree(tree_to_cgel, non_punct_tokens, '1', 'export', comment='')
+		block = io.StringIO(block)
+		cgel_tree = next(load_as_cgel(block))
+	except AssertionError:
+		_, _, tb = sys.exc_info()
+		traceback.print_tb(tb)
+		tb_info = traceback.extract_tb(tb)
+		# if we get this error, it means that ROOT has multiple children, which is not allowed by the CGEL parser.
+		# fallback strategy: place contents of ROOT node under a new node labeled 'Clause'
+		if tb_info[1].line == 'assert root is None':
+			tree_to_cgel.label = 'Clause'
+			tree_to_cgel = ParentedTree('ROOT', [tree_to_cgel])
+			block = writetree(tree_to_cgel, non_punct_tokens, '1', 'export', comment='')
+			block = io.StringIO(block)
+			cgel_tree = next(load_as_cgel(block))
+
+	cgel_tree_terminals = cgel_tree.terminals(gaps=True)
+
+	def unescape_ptree_tok(punct_token_list):
```

Comment:
slight naming thing: I would call the parameter `token_list` or `tokens` because it can be any sequence of tokens (only a few punctuation tokens will be affected)